File: security/moderation.py
"""Doubao Moderation内容安全层 — L5层统一安全检查.

五层安全体系:
- L1 输入净化: 基础输入清洗和格式校验
- L2 权限检查: 操作权限验证
- L3 架构约束: 层级依赖校验
- L4 工具沙箱: 工具执行隔离
- L5 内容审核: 发布前内容安全检查（本模块）

本模块实现L5内容审核，检查敏感词、隐私泄露、合规性三类风险。
"""

import re
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel

from flowforge.core.tracing import get_logger

logger = get_logger("security.moderation")


class ModerationLevel(str, Enum):
    """审核层级."""

    L1_INPUT_SANITIZATION = "L1"
    L2_PERMISSION_CHECK = "L2"
    L3_ARCHITECTURE_CONSTRAINT = "L3"
    L4_TOOL_SANDBOX = "L4"
    L5_CONTENT_MODERATION = "L5"


class ModerationResult(BaseModel):
    """审核结果."""

    safe: bool = True
    level: ModerationLevel = ModerationLevel.L5_CONTENT_MODERATION
    risk_tags: List[str] = []
    reason: str = ""
    confidence: float = 1.0


class ContentModerationChecker:
    """内容安全审核 — 发布前必经检查.

    检查维度:
    1. sensitive_words: 敏感词检测（政治/暴力/色情/诈骗等）
    2. privacy: 隐私泄露检测（手机号/身份证号/银行卡号）
    3. compliance: 合规性检测（虚假宣传等）

    用法:
        checker = ContentModerationChecker()
        result = await checker.check("待审核内容")
        if not result.safe:
            # 阻止发布或标记审核
    """

    # 敏感词分类（实际部署时应从配置文件加载）
    SENSITIVE_CATEGORIES: Dict[str, List[str]] = {
        "political": ["敏感词1", "敏感词2"],
        "violence": ["暴力", "凶杀"],
        "adult": ["色情", "裸体"],
        "fraud": ["诈骗", "传销"],
        "privacy": ["身份证号", "银行卡号"],
    }

    # 虚假宣传关键词
    FALSE_ADVERTISING_WORDS = ["100%有效", "绝对治愈", "包治百病"]

    async def check(
        self,
        content: str,
        check_types: Optional[List[str]] = None,
    ) -> ModerationResult:
        """执行内容安全检查.

        Args:
            content: 待审核的文本内容
            check_types: 检查类型列表，默认全部检查
                - "sensitive_words": 敏感词检测
                - "privacy": 隐私泄露检测
                - "compliance": 合规性检测

        Returns:
            ModerationResult 审核结果
        """
        types = check_types or ["sensitive_words", "privacy", "compliance"]
        risk_tags: List[str] = []

        if "sensitive_words" in types:
            tags = self._check_sensitive_words(content)
            risk_tags.extend(tags)

        if "privacy" in types:
            tags = self._check_privacy_leak(content)
            risk_tags.extend(tags)

        if "compliance" in types:
            tags = self._check_compliance(content)
            risk_tags.extend(tags)

        safe = len(risk_tags) == 0
        result = ModerationResult(
            safe=safe,
            risk_tags=risk_tags,
            reason="; ".join(risk_tags) if risk_tags else "内容安全检查通过",
            confidence=0.95 if safe else 0.85,
        )

        if not safe:
            logger.info(
                f"内容审核未通过: risk_tags={risk_tags}, "
                f"内容预览={content[:100]}"
            )

        return result

    def _check_sensitive_words(self, content: str) -> List[str]:
        """敏感词检查."""
        tags: List[str] = []
        for category, words in self.SENSITIVE_CATEGORIES.items():
            for word in words:
                if word in content:
                    tags.append(f"sensitive:{category}")
                    break  # 每个分类只标记一次
        return tags

    def _check_privacy_leak(self, content: str) -> List[str]:
        """隐私泄露检查."""
        tags: List[str] = []
        # 手机号
        if re.search(r"1[3-9]\d{9}", content):
            tags.append("privacy:phone_number")
        # 身份证号
        if re.search(r"\d{17}[\dXx]", content):
            tags.append("privacy:id_number")
        # 银行卡号（16-19位连续数字）
        if re.search(r"\d{16,19}", content):
            tags.append("privacy:bank_card")
        return tags

    def _check_compliance(self, content: str) -> List[str]:
        """合规性检查."""
        tags: List[str] = []
        # 检查虚假宣传
        if any(w in content for w in self.FALSE_ADVERTISING_WORDS):
            tags.append("compliance:false_advertising")
        return tags

    def add_sensitive_words(self, category: str, words: List[str]):
        """动态添加敏感词.

        Args:
            category: 敏感词分类
            words: 敏感词列表
        """
        if category not in self.SENSITIVE_CATEGORIES:
            self.SENSITIVE_CATEGORIES[category] = []
        self.SENSITIVE_CATEGORIES[category].extend(words)
        logger.info(f"添加敏感词: category={category}, count={len(words)}")

    def get_status(self) -> dict:
        """获取审核器状态."""
        return {
            "level": ModerationLevel.L5_CONTENT_MODERATION.value,
            "categories": list(self.SENSITIVE_CATEGORIES.keys()),
            "category_word_counts": {
                cat: len(words) for cat, words in self.SENSITIVE_CATEGORIES.items()
            },
        }

File: packages/plugins/guard/src/moderation.ts
/**
 * Content moderation — five-layer safety chain plus contract errors.
 *
 * Maps flowforge Python legacy core/moderation.py (contract only — the
 * external moderation API client itself stays host-injected) and
 * security/moderation.py (L5 publish-gate checker) (F25).
 *
 * Five-layer security architecture:
 * - L1 input sanitization: keyword matching
 * - L2 permission check: regex pattern rules
 * - L3 architecture constraint: general rules
 * - L4 tool sandbox / LLM review: injected provider
 * - L5 content moderation: injected platform provider
 *
 * Domain-specific rule sets are intentionally NOT bundled here; hosts
 * inject keywords/rules/providers appropriate for their deployment.
 */

import { createHash } from 'node:crypto'

// ---------------------------------------------------------------------------
// Contract errors (from core/moderation.py)
// ---------------------------------------------------------------------------

/** Moderation invocation failure (network/HTTP/parse errors). */
export class ModerationError extends Error {
  override readonly cause: unknown

  constructor(message: string, options: { cause?: unknown } = {}) {
    super(message)
    this.name = 'ModerationError'
    this.cause = options.cause
  }
}

/**
 * Business-level rejection: content hit a blocking risk label. Callers
 * should catch this and halt downstream flows (publish, persist, ...).
 */
export class ModerationBlockedError extends Error {
  readonly content: string
  readonly riskLabels: string[]

  constructor(content: string, riskLabels: string[]) {
    const preview = content.slice(0, 200)
    super(`Content blocked: risk_labels=${JSON.stringify(riskLabels)}, preview=${JSON.stringify(preview.slice(0, 80))}`)
    this.name = 'ModerationBlockedError'
    this.content = preview
    this.riskLabels = [...riskLabels]
  }
}

/** All moderation retries timed out; callers decide the fallback action. */
export class ModerationTimeoutError extends Error {
  readonly timeoutSeconds: number

  constructor(message: string, timeoutSeconds: number) {
    super(message)
    this.name = 'ModerationTimeoutError'
    this.timeoutSeconds = timeoutSeconds
  }
}

// ---------------------------------------------------------------------------
// Levels / severities / actions
// ---------------------------------------------------------------------------

export const MODERATION_LEVELS = ['L1', 'L2', 'L3', 'L4', 'L5'] as const
export type ModerationLevel = (typeof MODERATION_LEVELS)[number]

export type ModerationSeverity = 'none' | 'low' | 'medium' | 'high' | 'critical'
export type ModerationAction = 'allow' | 'warn' | 'block' | 'review'

export interface ModerationRule {
  name: string
  level: ModerationLevel
  category: string
  /** Literal substring (string) or regex test. */
  pattern: string | RegExp
  action: ModerationAction
  severity: ModerationSeverity
  enabled: boolean
}

export interface ModerationResult {
  passed: boolean
  level: ModerationLevel
  category: string
  severity: ModerationSeverity
  reason: string
  details: string[]
  action: ModerationAction
  cacheHit?: boolean
}

/** Injected provider outcome (L4 llm / L5 platform). */
export interface ModerationProviderOutcome {
  passed: boolean
  reason?: string
  category?: string
  severity?: ModerationSeverity
  details?: string[]
}

export interface ModerationProvider {
  check(content: string): Promise<ModerationProviderOutcome>
}

// ---------------------------------------------------------------------------
// Five-layer chain
// ---------------------------------------------------------------------------

export interface ContentModerationLayerOptions {
  /** L1: category → keyword list. */
  keywords?: Record<string, string[]>
  /** L2: regex pattern rules. */
  regexes?: ModerationRule[]
  /** L3: general rules. */
  rules?: ModerationRule[]
  /** L4: LLM-based review (host-injected). */
  llm?: ModerationProvider
  /** L5: platform moderation service (host-injected). */
  platform?: ModerationProvider
  /** Cache TTL for verdicts, seconds. */
  cacheTtlSeconds?: number
  /** Injectable clock (epoch seconds) for tests. */
  now?: () => number
}

interface CacheEntry {
  result: ModerationResult
  expiresAt: number
}

export class ContentModerationLayer {
  private readonly keywords: Record<string, string[]>
  private readonly regexes: ModerationRule[]
  private readonly rules: ModerationRule[]
  private readonly llm: ModerationProvider | undefined
  private readonly platform: ModerationProvider | undefined
  private readonly cacheTtlSeconds: number
  private readonly now: () => number
  private readonly cache = new Map<string, CacheEntry>()

  constructor(options: ContentModerationLayerOptions = {}) {
    this.keywords = options.keywords ?? {}
    this.regexes = options.regexes ?? []
    this.rules = options.rules ?? []
    this.llm = options.llm
    this.platform = options.platform
    this.cacheTtlSeconds = options.cacheTtlSeconds ?? 300
    this.now = options.now ?? (() => Date.now() / 1000)
  }

  /** Run the five-layer chain; any blocking layer short-circuits. */
  async moderate(content: string): Promise<ModerationResult> {
    const key = createHash('sha256').update(content).digest('hex')
    const cached = this.cache.get(key)
    if (cached && cached.expiresAt > this.now()) {
      return { ...cached.result, cacheHit: true }
    }

    const warnings: string[] = []

    // L1 keyword scan — one hit per category is enough to block.
    for (const [category, words] of Object.entries(this.keywords)) {
      const hit = words.find(word => content.includes(word))
      if (hit) {
        return this.finish(key, {
          passed: false,
          level: 'L1',
          category,
          severity: 'high',
          reason: `keyword '${hit}' matched`,
          details: [`keyword:${category}`],
          action: 'block',
        })
      }
    }

    // L2 regex rules.
    for (const rule of this.regexes) {
      const outcome = this.applyRule(content, rule, 'L2')
      if (outcome?.action === 'block') return this.finish(key, outcome)
      if (outcome) warnings.push(outcome.reason)
    }

    // L3 general rules.
    for (const rule of this.rules) {
      const outcome = this.applyRule(content, rule, 'L3')
      if (outcome?.action === 'block') return this.finish(key, outcome)
      if (outcome) warnings.push(outcome.reason)
    }

    // L4 LLM review.
    const llmOutcome = await this.runProvider(this.llm, content, 'L4')
    if (llmOutcome && !llmOutcome.passed) {
      return this.finish(key, this.providerResult(llmOutcome, 'L4'))
    }

    // L5 platform moderation.
    const platformOutcome = await this.runProvider(this.platform, content, 'L5')
    if (platformOutcome && !platformOutcome.passed) {
      return this.finish(key, this.providerResult(platformOutcome, 'L5'))
    }

    return this.finish(key, {
      passed: true,
      level: 'L5',
      category: '',
      severity: warnings.length > 0 ? 'low' : 'none',
      reason: warnings.length > 0 ? warnings.join('; ') : 'all layers passed',
      details: warnings,
      action: warnings.length > 0 ? 'warn' : 'allow',
    })
  }

  get cacheSize(): number {
    return this.cache.size
  }

  private applyRule(
    content: string,
    rule: ModerationRule,
    level: ModerationLevel,
  ): ModerationResult | null {
    if (!rule.enabled) return null
    const matched =
      typeof rule.pattern === 'string'
        ? content.includes(rule.pattern)
        : rule.pattern.test(content)
    if (!matched) return null
    return {
      passed: rule.action !== 'block',
      level,
      category: rule.category,
      severity: rule.severity,
      reason: `rule '${rule.name}' matched`,
      details: [`rule:${rule.name}`],
      action: rule.action,
    }
  }

  private async runProvider(
    provider: ModerationProvider | undefined,
    content: string,
    level: ModerationLevel,
  ): Promise<ModerationProviderOutcome | null> {
    if (!provider) return null
    try {
      return await provider.check(content)
    } catch (error) {
      throw new ModerationError(`Moderation provider ${level} failed`, { cause: error })
    }
  }

  private providerResult(
    outcome: ModerationProviderOutcome,
    level: ModerationLevel,
  ): ModerationResult {
    return {
      passed: false,
      level,
      category: outcome.category ?? '',
      severity: outcome.severity ?? 'high',
      reason: outcome.reason ?? `provider ${level} rejected content`,
      details: outcome.details ?? [],
      action: 'block',
    }
  }

  private finish(key: string, result: ModerationResult): ModerationResult {
    this.cache.set(key, { result, expiresAt: this.now() + this.cacheTtlSeconds })
    return result
  }
}

// ---------------------------------------------------------------------------
// L5 publish-gate checker (from security/moderation.py)
// ---------------------------------------------------------------------------

export interface ContentCheckResult {
  safe: boolean
  level: ModerationLevel
  riskTags: string[]
  reason: string
  confidence: number
}

export interface ContentModerationCheckerOptions {
  /** category → words; hosts load real word lists from config. */
  sensitiveCategories?: Record<string, string[]>
  /** False-advertising style compliance keywords. */
  complianceWords?: string[]
  /** Enable built-in PII (phone/id/bank-card) regex checks. */
  privacyChecks?: boolean
}

const PRIVACY_PATTERNS: Array<[string, RegExp]> = [
  ['privacy:phone_number', /1[3-9]\d{9}/],
  ['privacy:id_number', /\d{17}[\dXx]/],
  ['privacy:bank_card', /\d{16,19}/],
]

/**
 * Publish-time content check across three dimensions: sensitive words,
 * privacy leakage and compliance. Word lists are host-injected; only
 * neutral PII regexes ship by default.
 */
export class ContentModerationChecker {
  private readonly sensitiveCategories: Record<string, string[]>
  private readonly complianceWords: string[]
  private readonly privacyChecks: boolean

  constructor(options: ContentModerationCheckerOptions = {}) {
    // copy per-category lists so addSensitiveWords never mutates caller data
    this.sensitiveCategories = Object.fromEntries(
      Object.entries(options.sensitiveCategories ?? {}).map(([category, words]) => [
        category,
        [...words],
      ]),
    )
    this.complianceWords = options.complianceWords ?? []
    this.privacyChecks = options.privacyChecks ?? true
  }

  async check(content: string, checkTypes?: string[]): Promise<ContentCheckResult> {
    const types = checkTypes ?? ['sensitive_words', 'privacy', 'compliance']
    const riskTags: string[] = []

    if (types.includes('sensitive_words')) {
      riskTags.push(...this.checkSensitiveWords(content))
    }
    if (types.includes('privacy') && this.privacyChecks) {
      riskTags.push(...this.checkPrivacyLeak(content))
    }
    if (types.includes('compliance')) {
      riskTags.push(...this.checkCompliance(content))
    }

    const safe = riskTags.length === 0
    return {
      safe,
      level: 'L5',
      riskTags,
      reason: safe ? 'content safety check passed' : riskTags.join('; '),
      confidence: safe ? 0.95 : 0.85,
    }
  }

  private checkSensitiveWords(content: string): string[] {
    const tags: string[] = []
    for (const [category, words] of Object.entries(this.sensitiveCategories)) {
      // one tag per category
      if (words.some(word => content.includes(word))) tags.push(`sensitive:${category}`)
    }
    return tags
  }

  private checkPrivacyLeak(content: string): string[] {
    const tags: string[] = []
    for (const [tag, pattern] of PRIVACY_PATTERNS) {
      if (pattern.test(content)) tags.push(tag)
    }
    return tags
  }

  private checkCompliance(content: string): string[] {
    return this.complianceWords.some(word => content.includes(word))
      ? ['compliance:false_advertising']
      : []
  }

  addSensitiveWords(category: string, words: string[]): void {
    if (!this.sensitiveCategories[category]) this.sensitiveCategories[category] = []
    this.sensitiveCategories[category].push(...words)
  }

  getStatus(): Record<string, unknown> {
    return {
      level: 'L5',
      categories: Object.keys(this.sensitiveCategories),
      category_word_counts: Object.fromEntries(
        Object.entries(this.sensitiveCategories).map(([cat, words]) => [cat, words.length]),
      ),
    }
  }
}
